XMonad/Hooks/Modal.hs

```diff
@@ -0,0 +1,275 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+--------------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Hooks.Modal
+-- Description :  Implements true modality in xmonad key-bindings.
+-- Copyright   :  (c) 2018  L. S. Leary
+-- License     :  BSD3-style (see LICENSE)
+--
+-- Author      :  L. S. Leary
+-- Maintainer  :  Yecine Megdiche <yecine.megdiche@gmail.com>
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- This module implements modal keybindings for xmonad.
+--
+--------------------------------------------------------------------------------
+
+-- --< Imports & Exports >-- {{{
+module XMonad.Hooks.Modal
+  (
+ -- * Usage
+ -- $Usage
+    modal
+  , mode'
+  , mode
+  , Mode
+  , setMode
+  , exitMode
+ -- ** Provided Modes
+ -- $ProvidedModes
+  , noModModeLabel
+  , noModMode
+  , floatModeLabel
+  , floatMode
+  , overlayedFloatModeLabel
+  , overlayedFloatMode
+  , floatMap
+  , overlay
+ -- ** Logger
+  , logMode
+  ) where
+
+-- core
+import           XMonad
+
+-- base
+import           Data.Bits                      ( (.&.)
+                                                , complement
+                                                )
+import           Data.List
+import qualified Data.Map.Strict               as M
+import           XMonad.Actions.FloatKeys       ( keysMoveWindow
+                                                , keysResizeWindow
+                                                )
+import           XMonad.Prelude
+import qualified XMonad.Util.ExtensibleConf    as XC
+-- contrib
+import qualified XMonad.Util.ExtensibleState   as XS
+import           XMonad.Util.Grab
+import           XMonad.Util.Loggers
+
+-- }}}
+
+-- Original Draft By L.S.Leary : https://gist.github.com/LSLeary/6741b0572d62db3f0cea8e6618141b2f
+
+-- --< Usage >-- {{{
+
+-- $Usage
+--
+-- This module provides modal keybindings in xmonad. You can think of modes as
+-- submaps from 'XMonad.Actions.Submap', but after each action you execute, you
+-- land back into the submap until you explicitly exit the submap.
```

Comment:
Now that I'm a bit more awake, "into the submap" seems like it should be "in the submap". (I focused more on the code in my first review and pretty much ignored the comments.)